Autodesk/synthesis
# Input System Touch Controls

## Description
Creating touch controls joysticks for mobile devices. To enable the touch controls, you must be on a device with touch capabilities. When this happens, the Enable touch controls button will appear in the MainHUD.

<img width="356" alt="Screenshot 2024-08-23 at 2 48 12 PM" src="https://github.com/user-attachments/assets/8b5e6dbf-941d-4a6e-8d5d-a00d7e1480fd">

After enabling the touch controls the controls themselves will appear on screen.

<img width="1119" alt="Screenshot 2024-08-23 at 2 49 06 PM" src="https://github.com/user-attachments/assets/28d557f2-8d06-46b6-a76b-aec3f3961584">

To use them, you must configure a inputscheme or use the preset "Brandon"

### Notes
- In the future we will create a queue of the different actions on the top of the screen that will appear as they are available to execute. For example, intake when there is an object in the intakesensor, eject when needed, and place assembly. These will be based off the design of the place assembly button.
- When configuring the inputscheme, the useGamepad and useTouchControls checkboxes can be selected at the same time. This is a bug. To fix this, we will need to change the implementation of the Checkbox to utilize the checked={} prop instead of default value.
- When creating a new configuration and enabling touch controls, the different joint configurations appear (which causes a bug of joints moving as the robot moves). To fix this, we should create a "add joint configuration" button when creating a new configuration.
- In the future, on the top left of the screen there will be sliders for joints similar to the queue that will appear only if there is a configuration for a joint.


### Testing

To test this on a mobile device:
1. Enable SSL by going to vite.config.js and enabling the boolean to be true
2. run `bun run host` or `npm run host` to expose it to the network
3. Open the URL (not the localhost url) on your mobile device that is connected to the same network.
> **Note:** Ensure you are not on a wifi network with a firewall that blocks external access (such as the Autodesk network). Alternatively, you can use a hotspot.

[JIRA Issue](https://jira.autodesk.com/browse/AARD-1847)
